File: Kalorie/Kalorie/Core/UseCases/SaveFoodConsumedUseCase.swift
//
//  SaveFoodConsumedUseCase.swift
//  Kalorie
//
//  Created by Josef Antoni on 01.07.2026.
//

import Foundation

protocol SaveFoodConsumedUseCaseProtocol {
    func callAsFunction(_ item: FoodItemDomain, grams: Double, date: Date) async throws
}

struct SaveFoodConsumedUseCase: SaveFoodConsumedUseCaseProtocol {

    // MARK: - Properties

    private let dataProvider: any FirestoreDataProviderProtocol
    private let authProvider: any AuthProviderProtocol

    // MARK: - Init

    init(dataProvider: any FirestoreDataProviderProtocol, authProvider: any AuthProviderProtocol) {
        self.dataProvider = dataProvider
        self.authProvider = authProvider
    }

    // MARK: - Functions

    func callAsFunction(_ item: FoodItemDomain, grams: Double, date: Date) async throws {
        guard let userId = authProvider.userId else { throw AuthError.notAuthenticated }
        let scaled = item.scaled(toGrams: grams)
        let dto = FoodConsumedDTO(
            id: UUID().uuidString,
            foodItemId: item.id,
            foodItemKind: item.kind,
            czName: item.czName,
            engName: item.engName,
            weight: grams,
            date: date.timeIntervalSince1970,
            calories: scaled.calories,
            protein: scaled.protein,
            carbohydrate: scaled.carbohydrate,
            carbohydrateSugar: scaled.carbohydrateSugar,
            fat: scaled.fat,
            fatUnsaturated: scaled.fatUnsaturated,
            fiber: scaled.fiber,
            salt: scaled.salt
        )
        try await dataProvider.setAsync(dto, id: dto.id, in: Constants.Firestore.foodConsumed(userId: userId))
    }
}

#if DEBUG
struct SaveFoodConsumedUseCaseFake: SaveFoodConsumedUseCaseProtocol {

    // MARK: - Properties

    var shouldThrow = false

    // MARK: - Functions

    func callAsFunction(_ item: FoodItemDomain, grams: Double, date: Date) async throws {
        if shouldThrow { throw URLError(.unknown) }
    }
}
#endif

File: Android/app/src/main/java/antoni/kalorie/core/usecases/SaveFoodConsumedUseCase.kt
package antoni.kalorie.core.usecases

import antoni.kalorie.core.auth.AuthError
import antoni.kalorie.core.auth.AuthProviderProtocol
import antoni.kalorie.core.models.FoodItemDomain
import antoni.kalorie.core.models.scaled
import antoni.kalorie.core.networking.FirestoreDataProviderProtocol
import antoni.kalorie.core.networking.FoodConsumedDTO
import antoni.kalorie.core.networking.setAsync
import antoni.kalorie.core.utils.Constants
import antoni.kalorie.core.utils.epochSecondsAsDouble
import java.time.Instant
import java.util.UUID

interface SaveFoodConsumedUseCaseProtocol {
    suspend operator fun invoke(item: FoodItemDomain, grams: Double, date: Instant, mealTypeId: String?)
}

class SaveFoodConsumedUseCase(
    private val dataProvider: FirestoreDataProviderProtocol,
    private val authProvider: AuthProviderProtocol,
) : SaveFoodConsumedUseCaseProtocol {

    // MARK: - Functions

    override suspend fun invoke(item: FoodItemDomain, grams: Double, date: Instant, mealTypeId: String?) {
        val userId = authProvider.userId ?: throw AuthError.NotAuthenticated
        val scaled = item.scaled(toGrams = grams)
        val dto = FoodConsumedDTO(
            id = UUID.randomUUID().toString().uppercase(),
            foodItemId = item.id,
            foodItemKind = item.kind,
            czName = item.czName,
            engName = item.engName,
            weight = grams,
            date = date.epochSecondsAsDouble(),
            calories = scaled.calories,
            caloriesPerHundredGrams = item.caloriesPerHundredGrams,
            energyKJ = scaled.energyKJ,
            protein = scaled.protein,
            carbohydrate = scaled.carbohydrate,
            carbohydrateSugar = scaled.carbohydrateSugar,
            fat = scaled.fat,
            fatSaturated = scaled.fatSaturated,
            fatUnsaturated = scaled.fatUnsaturated,
            fiber = scaled.fiber,
            salt = scaled.salt,
            mealTypeId = mealTypeId,
            measureUnit = item.measure.rawValue,
        )
        dataProvider.setAsync(dto, id = dto.id, inCollection = Constants.Firestore.foodConsumed(userId))
    }
}
